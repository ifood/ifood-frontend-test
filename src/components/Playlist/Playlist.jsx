import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { getPlaylists } from '../../store/actions';

import { InnerContainer } from '../style/container/Container';

const Playlist = () => {
    const userToken = useSelector(({ spotify }) => spotify.userToken);
    const playlists = useSelector(({ spotify }) => spotify.playlists);
    const dispatch = useDispatch();

    useEffect(() => {
        dispatch(getPlaylists({ userToken }));
    }, [dispatch, userToken]);

    return playlists ? (
        <InnerContainer>
            {playlists.map((list) => (
                <li key={list.uri}>{list.name}</li>
            ))}
        </InnerContainer>
    ) : (
        <InnerContainer>Teste2</InnerContainer>
    );
};

export default Playlist;
