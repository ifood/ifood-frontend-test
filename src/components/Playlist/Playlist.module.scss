@import 'scss/components.scss';

/* */

$loading-placeholder-color: $md-grey-300;
$loading-placeholder-color-dark: $md-grey-400;

$theme-dark-loading-placeholder-color: darken($md-grey-800, 7.5%);
$theme-dark-loading-placeholder-color-dark: $md-grey-800;

$theme-high-contrast-loading-placeholder-color: black;
$theme-high-contrast-loading-placeholder-color-dark: rgba(black, .75);

$theme-dark-high-contrast-loading-placeholder-color: black;
$theme-dark-high-contrast-loading-placeholder-color-dark: rgba(white, .1);

/* */

%loadingPlaceholder {

    animation: loadingPlaceholder 3s infinite;

    @keyframes loadingPlaceholder {

        0%, 75%, 100% {

            background: $loading-placeholder-color;

        }

        25% {

            background: $loading-placeholder-color-dark;

        }

    }

}

%loadingPlaceholderDark {

    animation: loadingPlaceholderDark 3s infinite;

    @keyframes loadingPlaceholderDark {

        0%, 75%, 100% {

            background: $theme-dark-loading-placeholder-color;

        }

        25% {

            background: $theme-dark-loading-placeholder-color-dark;

        }

    }

}

%loadingPlaceholderHighContrast {

    animation: loadingPlaceholderHighContrast 3s infinite;

    @keyframes loadingPlaceholderHighContrast {

        0%, 75%, 100% {

            background: $theme-high-contrast-loading-placeholder-color;

        }

        25% {

            background: $theme-high-contrast-loading-placeholder-color-dark;

        }

    }

}

%loadingPlaceholderDarkHighContrast {

    animation: loadingPlaceholderDarkHighContrast 3s infinite;

    @keyframes loadingPlaceholderDarkHighContrast {

        0%, 75%, 100% {

            background: $theme-dark-high-contrast-loading-placeholder-color;

        }

        25% {

            background: $theme-dark-high-contrast-loading-placeholder-color-dark;

        }

    }

}

:global {

    .playlist-slider {

        &-enter {

            transform: translateX(-100%);
            // margin-right: -250px !important;

        }

        &-enter-active {

            transform: translateX(0);
            // margin-right: 0 !important;

            transition: .3s ease;
            transition-property: transform, margin-right;

        }

        &-exit-active {

            transform: translateX(-100%);
            // margin-right: -250px !important;

            transition: .3s ease;
            transition-property: transform, margin-right;

        }

    }

}

.Playlist {

    &Title {

        margin: 0;
        margin-bottom: $theme-padding * 2;

        font-family: 'Barlow', sans-serif;
        @include font-sizer(28);
        line-height: 1;
        color: $md-red-500;

        :global(.theme--light--high-contrast) & {

            color: black;

        }

        :global(.theme--dark--high-contrast) & {

            color: white

        }

        @media(max-width: $break-point-md){

            margin-bottom: $theme-padding * 1.1;

            @include font-sizer(24);

        }

        &Loading {

            width: 30%;

            background: $loading-placeholder-color;

            border-radius: 6px;

            @extend %loadingPlaceholder;

            :global(.theme--dark) & {

                background: $theme-dark-loading-placeholder-color;

                @extend %loadingPlaceholderDark;

            }

            :global(.theme--light--high-contrast) & {

                background: $theme-high-contrast-loading-placeholder-color;

                box-shadow: inset 0 0 0 2px black;

                @extend %loadingPlaceholderHighContrast;

            }

            :global(.theme--dark--high-contrast) & {

                background: $theme-dark-high-contrast-loading-placeholder-color;

                box-shadow: inset 0 0 0 2px white;

                @extend %loadingPlaceholderDarkHighContrast;

            }

        }

    }

    &List {

        &:not(:last-of-type){

            margin-bottom: $theme-padding * 5;

            @media(max-width: $break-point-md){

                margin-bottom: $theme-padding * 2;

            }

        }

        &Navigation {

            $size: 24px;

            width: $size;
            height: $size;

            padding: 0;

            background: rgba(black, .4);

            border: 0;
            border-radius: $size / 2;

            outline: none !important;

            cursor: pointer;

            transition: .2s ease;
            transition-property: opacity, transform;

            :global(.theme--dark) & {

                background: rgba(white, .5);

            }

            :global(.theme--light--high-contrast) & {

                background: black;

            }

            :global(.theme--dark--high-contrast) & {

                background: white;

            }

            &:not([disabled]):active {

                transform: scale(0.9)

            }

            &:not(:last-of-type){

                margin-right: $theme-padding / 2;

            }

            &[disabled]{

                opacity: 0.25;

                cursor: default;

            }

            svg {

                width: inherit;
                height: inherit;

                fill: white;

                :global(.theme--dark--high-contrast) & {

                    fill: black;

                }

            }

        }

        &Row {

            margin-left: -($theme-padding * 2);
            margin-top: -$theme-padding;
            padding: 0;
            padding-top: $theme-padding;
            padding-left: $theme-padding;

            list-style: none;

            display: flex;

            transition-timing-function: cubic-bezier(0.165, 0.84, 0.44, 1);
            transition-property: transform;

        }

        &List {

            padding-left: $theme-padding;
            padding-right: $theme-padding;

            display: inline-block;

            backface-visibility: hidden;
            perspective: 1000;

        }

        &Cover {

            $size-xl: 220px;
            $size-lg: 180px;
            $size-md: 130px;

            width: $size-xl;
            height: $size-xl;

            margin-bottom: $theme-padding;

            border-radius: 8px;

            :global(.theme--light--high-contrast) & {

                border: 2px solid black;

            }

            :global(.theme--dark--high-contrast) & {

                border: 2px solid white;

            }

            // box-shadow: 0 20px 20px rgba(black, .1);

            @media(max-width: $break-point-lg){

                width: $size-lg;
                height: $size-lg;

            }

            @media(max-width: $break-point-md){

                width: $size-md;
                height: $size-md;

            }

            &Loading {

                background: $loading-placeholder-color;

                @extend %loadingPlaceholder;

                :global(.theme--dark) & {

                    background: $theme-dark-loading-placeholder-color;

                    @extend %loadingPlaceholderDark;

                }

                :global(.theme--light--high-contrast) & {

                    background: $theme-high-contrast-loading-placeholder-color;

                    @extend %loadingPlaceholderHighContrast;

                }


                :global(.theme--dark--high-contrast) & {

                    background: $theme-dark-high-contrast-loading-placeholder-color;

                    @extend %loadingPlaceholderDarkHighContrast;

                }

            }

        }

        &Title {

            font-family: 'Barlow', sans-serif;
            @include font-sizer(16);
            line-height: 1;
            font-weight: 600;
            color: $md-grey-800;

            :global(.theme--dark) & {

                color: white;

            }

            :global(.theme--light--high-contrast) & {

                color: black;

                text-decoration: underline;

            }

            :global(.theme--dark--high-contrast) & {

                color: white;

                text-decoration: underline;

            }

            &Loading {

                width: 80%;

                background: $loading-placeholder-color;

                display: inline-block;

                border-radius: 8px;

                @extend %loadingPlaceholder;

                :global(.theme--dark) & {

                    background: $theme-dark-loading-placeholder-color;

                    @extend %loadingPlaceholderDark;

                }

                :global(.theme--light--high-contrast) & {

                    text-decoration: none;

                    background: $theme-high-contrast-loading-placeholder-color;

                    box-shadow: inset 0 0 0 2px black;

                    @extend %loadingPlaceholderHighContrast;

                }

                :global(.theme--dark--high-contrast) & {

                    text-decoration: none;

                    background: $theme-dark-high-contrast-loading-placeholder-color;

                    box-shadow: inset 0 0 0 2px white;

                    @extend %loadingPlaceholderDarkHighContrast;

                }

            }

        }

        &Owner {

            font-family: 'Barlow', sans-serif;
            @include font-sizer(12);
            font-weight: 400;
            color: rgba(black, .5);

            :global(.theme--dark) & {

                color: rgba(white, .5);

            }

            :global(.theme--light--high-contrast) & {

                color: black;

                text-decoration: underline;

            }

            :global(.theme--dark--high-contrast) & {

                color: white;

                text-decoration: underline;

            }

            &Loading {

                width: 60%;

                background: $loading-placeholder-color;

                display: inline-block;

                border-radius: 8px;

                @extend %loadingPlaceholder;

                :global(.theme--dark) & {

                    background: $theme-dark-loading-placeholder-color;

                    @extend %loadingPlaceholderDark;

                }

                :global(.theme--light--high-contrast) & {

                    text-decoration: none;

                    background: $theme-high-contrast-loading-placeholder-color;

                    box-shadow: inset 0 0 0 2px black;

                    @extend %loadingPlaceholderHighContrast;

                }

                :global(.theme--dark--high-contrast) & {

                    text-decoration: none;

                    background: $theme-dark-high-contrast-loading-placeholder-color;

                    box-shadow: inset 0 0 0 2px white;

                    @extend %loadingPlaceholderDarkHighContrast;

                }

            }

        }

    }

}
