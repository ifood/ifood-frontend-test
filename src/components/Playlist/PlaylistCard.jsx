import React from 'react';

import styled from 'styled-components';
import { randomColors, mainColor, white, black } from '../style/colors';

export const Card = styled.div`
    position: relative;
    width: 200px;
    height: 200px;
    display: flex;
    flex: 0 1 auto;
    align-content: flex-start;
    padding: 20px;
    flex-flow: column nowrap;
    justify-content: space-around;
    background: ${(props) => props.backgroundColor || mainColor};
    margin-bottom: 20px;

    div {
        flex-grow: 1;
        margin-right: 5px;

        &:last-child {
            margin-right: 0;
        }
    }

    @media (max-width: 1024px) {
        width: 100%;

        div {
            margin-right: 0;
        }
    }
`;

export const CardImage = styled.div`
    width: 80px;
    height: 80px;
    background: url(${(props) =>
        props.backgroundImage || 'https://dummyimage.com/80x80/000/fff&text=x'});
    background-repeat: no-repeat;
    background-size: 80px 80px;
    position: absolute;
    top: 20px;
    left: 20px;
    opacity: 0.6;

    @media (max-width: 1024px) {
        opacity: 1;
    }
`;

export const Name = styled.h1`
    font-family: 'Oswald', sans-serif;
    font-size: 36px;
    font-weight: 700;
    text-transform: uppercase;
    text-align: right;
    color: ${white};
`;

export const Author = styled.h1`
    font-family: 'Oswald', sans-serif;
    font-size: 16px;
    text-align: right;
    color: ${black};
`;

export const Description = styled.p`
    margin-top: 5px;
    font-family: 'Noto Sans JP', sans-serif;
    font-size: 16px;
    color: ${white};
`;

const PlayListCard = ({ name, description, owner, images }) => {
    const { display_name: displayName, href } = owner;
    const backgroundImage = images[0].url;
    const randomNumber = Math.floor(Math.random() * 10);

    return (
        <Card backgroundColor={randomColors[randomNumber]}>
            <CardImage backgroundImage={backgroundImage} />
            <Name>{name}</Name> <Author>by {displayName}</Author>
            <Description>{description}</Description>
        </Card>
    );
};

export default PlayListCard;
