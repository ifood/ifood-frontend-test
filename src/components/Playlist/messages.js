import { defineMessage } from 'react-intl';

const messages = {
  tracks: {
    id: 'playlist.tracks',
    defaultMessage: 'tracks',
  },
};

export default defineMessage(messages);
