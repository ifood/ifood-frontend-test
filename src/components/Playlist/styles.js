import styled from 'styled-components';
import { gray, lightGray } from '../../styles/colors';

export const Container = styled.li`
  .info-playlist {
    margin-left: 10px;
  }

  @media only screen and (min-width: 992px) {
    .col-md-2 {
      max-width: 12% !important;
    }
  }

  @media only screen and (max-width: 767px) {
    .name-info-playlist {
      font-size: 24px;
    }
  }
`;

export const Info = styled.p`
  margin-top: 8px;
  font-size: 12px;
  color: ${gray};
  margin-bottom: 0px;
`;

export const Name = styled.p`
  color: ${gray};
  font-weight: bold;
  font-size: 36px;
  margin: 0px;
`;

export const Owner = styled.p`
  color: ${lightGray};
  font-weight: bold;
  font-size: 12px;

  span {
    color: ${gray};
  }
`;
