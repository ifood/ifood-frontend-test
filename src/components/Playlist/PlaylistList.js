import React from 'react'
import _lang from 'lodash/lang'

/* */

import Cover from 'components/Cover/Cover'
import Icon from 'components/Icon/Icon'

/* */

import styles from './Playlist.module.scss'

/* */

class PlaylistList extends React.Component {

    constructor(props){

        super(props)

        this.state = {

            index: 0,
            translate: 0,
            duration: '0s'

        }

    }

    /* */

    async componentDidUpdate(prevProps, prevState){

        if(!_lang.isEqual(prevProps.data, this.props.data)){

            this.setState({ index: 0 })

        }

    }

    /* */

    getSlideStepWidth(){

        if(this.props.viewport < 768){

            return 160

        } else if(this.props.viewport > 768 && this.props.viewport <= 992){

            return 210

        } else {

            return 250

        }

    }

    getItemsOffset(){

        return Math.ceil(this.props.viewport / this.getSlideStepWidth()) + 2

    }

    getItems(){

        if(this.props.data.loading){

            return [...new Array(this.getItemsOffset()).keys()]

        } else {

            return this.props.data.items

        }

    }

    /* */

    handleNavigation(direction){

        let duration = 600

        /* */

        if(direction === 'prev'){

            if(this.state.index) this.setState({ index: this.state.index - 1 })

        } else {

            if(this.state.index < this.props.data.items.length - 1) this.setState({ index: this.state.index + 1 })

        }

        /* */

        window.requestAnimationFrame(() => {

            this.setState({

                translate: direction === 'prev' ? -this.getSlideStepWidth() : this.getSlideStepWidth(),
                duration : '0s'

            })

            setTimeout(() => {

                this.setState({

                    translate: 0,
                    duration : `${(duration / 1000)}s`

                })

                setTimeout(() => {

                    this.setState({ duration : '0s' })

                }, duration)

            }, 1)

        })

    }

    /* */

    render(){

        const getItems = this.getItems()
        const getItemsOffset = this.getItemsOffset()
        const getSlideStepWidth = this.getSlideStepWidth()

        /* */

        return (

            <div className={ styles.PlaylistList }>

                <div className="row no-gutters">

                    <div className="col">

                        <h1 className={

                            [

                                styles.PlaylistTitle,
                                this.props.data.loading && styles.PlaylistTitleLoading

                            ].filter(Boolean).join(' ')

                        }>{

                            !this.props.data.loading ? this.props.data.message : <>&nbsp;</>

                    }</h1>

                    </div>

                    {

                        !this.props.data.loading && (

                            <div className="col-auto pr-4 pr-md-5">

                                <button

                                className={ styles.PlaylistListNavigation }

                                disabled={ !this.state.index }
                                onClick={ () => this.handleNavigation('prev') }

                                title="Anterior"

                                aria-label="Anterior"

                                >

                                    <Icon glyph="arrow_left" />

                                </button>

                                <button

                                className={ styles.PlaylistListNavigation }

                                disabled={ this.state.index === this.props.data.items.length - 1 }
                                onClick={ () => this.handleNavigation('next') }

                                title="Próximo"

                                aria-label="Próximo"

                                >

                                    <Icon glyph="arrow_right" />

                                </button>

                            </div>

                        )

                    }

                </div>

                {

                getItems.length ?

                <div className={ styles.PlaylistSlider }>

                    <ul className={ styles.PlaylistListRow } style={{

                        transform : `translate3d(${this.state.translate - getSlideStepWidth}px, 0, 0)`,
                        transitionDuration : this.state.duration

                    }}

                    >

                        {

                            getItems.slice(this.state.index, this.state.index + getItemsOffset).map((val, index) => {

                                return (

                                    <li className={ styles.PlaylistListList } key={ val.id || index }>

                                        <div className={

                                            [

                                                styles.PlaylistListCover,
                                                this.props.data.loading && styles.PlaylistListCoverLoading

                                            ].filter(Boolean).join(' ')

                                        } style={{

                                            animationDelay : `${(0.4 * index)}s`

                                        }} role="img" aria-label={ `Imagem de capa ${(val.type === 'album' ? 'do álbum' : 'da playlist')} "${val.title}" de ${val.owner}` }>

                                            {

                                                (val.img && val.uri) ? (

                                                    <Cover data={ val } />

                                                ) : null

                                            }

                                        </div>

                                        <div className={

                                            [

                                                styles.PlaylistListTitle,
                                                this.props.data.loading && styles.PlaylistListTitleLoading

                                            ].filter(Boolean).join(' ')

                                        }

                                        style={{

                                           animationDelay : `${(0.4 * index)}s`

                                       }}

                                        >{ !this.props.data.loading ? val.title : <>&nbsp;</> }</div>

                                        {

                                            (val.owner || this.props.data.loading) && (

                                                <div className={

                                                    [

                                                        styles.PlaylistListOwner,
                                                        this.props.data.loading && styles.PlaylistListOwnerLoading

                                                    ].join(' ')

                                                }

                                                style={{

                                                   animationDelay : `${(0.4 * index)}s`

                                               }}

                                                >{ !this.props.data.loading ? `de ${val.owner}` : <>&nbsp;</> }</div>

                                            )

                                        }

                                    </li>

                                )

                            })

                        }

                    </ul>

                </div>

                : null

                }

            </div>

        )

    }

}

export default PlaylistList
