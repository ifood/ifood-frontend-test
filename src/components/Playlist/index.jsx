/* eslint-disable react/prefer-stateless-function */
import { Component } from 'react';

import { Container, Info, Name, Owner } from './styles';

class Playlist extends Component {
  render() {
    const { playlist } = this.props;

    return (
      <Container className="row featurette">
        <div className="col-md-2">
          <a
            className="url-info-playlist"
            href={playlist.external_urls.spotify}
            target="blank"
          >
            <img height={100} src={playlist.images[0].url} alt="SpotiFood" />{' '}
          </a>
        </div>
        <div className="col-md-10">
          <Info>PLAYLIST</Info>
          <Name>{playlist.name}</Name>
          <Owner>
            Created by <span>{playlist.owner.display_name}</span> -{' '}
            {playlist.tracks.total} songs
          </Owner>
        </div>
      </Container>
    );
  }
}

export default Playlist;
