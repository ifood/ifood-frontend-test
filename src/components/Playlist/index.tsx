import React from 'react';
import { usePlaylist } from '../../hooks/playlists';

import * as S from './styles';

const Playlist: React.FC = () => {
  const { playlists } = usePlaylist();

  return (
    <S.PlaylistContainer>
      {playlists.length > 0 ? (
        playlists.map(playlist => (
          <a
            href={playlist.external_urls.spotify}
            target="__blank"
            key={playlist.name}
          >
            <S.PlaylistCard>
              <div className="card-inner">
                <S.CardFront image={playlist.images[0].url} />
                <S.CardBack>
                  <h2>{playlist.name}</h2>
                  <p>{playlist.description}</p>
                </S.CardBack>
              </div>
            </S.PlaylistCard>
          </a>
        ))
      ) : (
        <h1>No playlist match with this search</h1>
      )}
    </S.PlaylistContainer>
  );
};

export default Playlist;
