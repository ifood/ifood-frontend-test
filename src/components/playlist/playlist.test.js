import React from 'react';
import Playlist from './playlist';
import { shallow } from 'enzyme';

// Mounting - Full DOM rendering including child components
// Shallow - Renders only the single component, not including its children,
// this is useful to isolate the component for pure unit testing.
// Render - Renders to static HTML, including children

it('should render correctly with props', () => {
    const playlist = {
        "name":"Sweet Soul Chillout",
        "imageUrl":"https://pl.scdn.co/images/pl/default/dd52dbcce4e10a2feb519e9c282b4d3b0dcc7442",
        "numberOfTracks": "36",
        "owner":"Spotify"};

    const component = shallow(<Playlist playlist={playlist}/>);
    expect(component).toMatchSnapshot();
});

