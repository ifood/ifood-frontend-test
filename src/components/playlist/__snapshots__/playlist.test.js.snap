// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with props 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#808080",
      "color": "#000",
      "display": "inline-block",
      "fontFamily": "Papyrus",
      "padding": "10px",
      "width": "25%",
    }
  }
>
  <img
    alt="Image URL index: undefined"
    src="https://pl.scdn.co/images/pl/default/dd52dbcce4e10a2feb519e9c282b4d3b0dcc7442"
    style={
      Object {
        "width": "60px",
      }
    }
  />
  <h3
    style={
      Object {
        "fontWeight": "bold",
      }
    }
  >
    Sweet Soul Chillout
  </h3>
  <ul
    style={
      Object {
        "marginTop": "10px",
      }
    }
  >
    <li
      style={
        Object {
          "paddingTop": "2px",
        }
      }
    >
      <span
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Owner: 
      </span>
      Spotify
    </li>
    <li
      style={
        Object {
          "paddingTop": "2px",
        }
      }
    >
      <span
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Total of tracks: 
      </span>
      36
    </li>
  </ul>
</div>
`;
